src/coreclr/vm/jitinterface.cpp

```diff
@@ -1636,6 +1645,60 @@ uint32_t CEEInfo::getThreadLocalFieldInfo (CORINFO_FIELD_HANDLE  field, bool isG
     return typeIndex;
 }
 
+#ifndef _MSC_VER
+
+extern "C" void* JIT_GetThreadStaticsBaseOffset();
+
+#ifdef HOST_AMD64
+
+void* getThreadStaticDescriptor(uint8_t* p)
+{
+#ifdef TARGET_OSX
+    if (!(p[0] == 0x48 && p[1] == 0x8d && p[2] == 0x3d))
+    {
+        // The optimization is disabled if coreclr is not compiled in .so format.
+        return 0;
+    }
+
+    // At this point, `p` contains the instruction pointer and is pointing to the above opcodes.
+    // These opcodes are patched by the dynamic linker.
+    // Move beyond the opcodes that we have already checked above.
+    p += 3;
+#else
+    if (!(p[0] == 0x66 && p[1] == 0x48 && p[2] == 0x8d && p[3] == 0x3d))
+    {
+        // The optimization is disabled if coreclr is not compiled in .so format.
+        return 0;
+    }
+
+    // At this point, `p` contains the instruction pointer and is pointing to the above opcodes.
+    // These opcodes are patched by the dynamic linker.
+    // Move beyond the opcodes that we have already checked above.
+    p += 4;
+#endif
+
+    // The descriptor address is located at *p at this point. Read that and add
+    // it to the instruction pointer to locate the address of `ti` that will be used
+    // to pass to __tls_get_addr during execution.
+    // (p + 4) below skips the descriptor address bytes embedded in the instruction and
+    // add it to the `instruction pointer` to find out the address.
+    return *(uint32_t*)p + (p + 4);
+}
+
+void* getTlsIndexObjectAddress()
+{
+    uint8_t* p = reinterpret_cast<uint8_t*>(&JIT_GetThreadStaticsBaseOffset);
+    return getThreadStaticDescriptor(p);
+}
+
+#elif HOST_ARM64
+
+uint64_t getTlsIndexObjectAddress()
```

Comment:
This does not return tls_index object address. The name should match what it returns.